runtime/doc/quickfix.txt

```diff
@@ -1335,6 +1335,59 @@ g:javac_makeprg_params variable.  For example: >
 
 	let g:javac_makeprg_params = "-Xlint:all -encoding utf-8"
 <
+SPOTBUGS						*compiler-spotbugs*
+
+SpotBugs is a static analysis tool that can be used to find bugs in Java.
+It scans the Java bytecode of all classes in the currently open buffer.
+
+Commonly used compiler options can be added to 'makeprg' by setting the
+b/g:spotbugs_makeprg_params variable.  For example: >
+
+	let b:spotbugs_makeprg_params = "-longBugCodes -effort:max -low"
+
+The global default is "-workHard -experimental".
+
+By default, the class files are searched in the directory where the source
+files are placed.  However, typical Java projects use distinct directories
+for source files and class files.  To make both known to SpotBugs, assign
+their paths (relative to the current directory) to the following two
+properties (using the example of a typical Maven project): >
+
+	let g:spotbugs_properties = {'sourceDirName': 'src/main/java',
+				   \ 'classDirName':  'target/classes'}
+
+Pre- and post-compiler actions, to first refresh the class files for the
+current buffer, and then run `spotbugs` on them, can be enabled by adding
+to your |vimrc| (before any |FileType| event): >
+
+	let g:spotbugs_properties = {
+		\ 'PreCompilerAction':  function('CompilerJavacMake'),
+		\ 'PostCompilerAction': function('SpotBugsMake'),
+	\}
+
+	function! CompilerJavacMake() abort
+		if exists('b:spotbugs_class_files')
+			for pathname in b:spotbugs_class_files
+				call delete(eval(pathname))
+			endfor
+
+			let b:spotbugs_class_files = []
+		endif
+
+  		compiler javac
+  		exe 'make -Xlint -Xdiags:verbose'
+			\ '-d' shellescape(get(g:, 'spotbugs_properties.classDirName', '.'), 1) '%:S'
+	endfunction
+
+	function! SpotBugsMake() abort
+		make %:S
+	endfunction
```

Comment:
The declaration order should be reversed (`E700`): first go  
function declarations, then goes the dictionary that makes  
Funcrefs out of them.